backend/lib/endpoints/posts.js

```diff
@@ -127,19 +216,102 @@ async function routes(app) {
 
   app.patch(
     "/:postId",
-    { preValidation: [app.authenticate], schema: updatePostSchema },
+    {
+      preValidation: [app.authenticate],
+      schema: updatePostSchema,
+    },
     async (req, reply) => {
-      // todo: make sure user can only update their own post
+      // TODO: get userId from jwt
+      // userId = ?
+      // user = User.findById(userId);
+      var user = await User.findById("5ea6900c0e0419d4cb123611");
+      if (user === null) return new httpErrors.Unauthorized();
+
       const post = await Post.findById(req.params.postId);
-      if (post === null) {
-        return reply.send(new httpErrors.NotFound());
-      }
-      Object.keys(req.body).forEach((key) => {
-        if (post[key] && post[key] !== req.body[key]) {
-          post[key] = req.body[key];
+      if (!post) return new httpErrors.NotFound();
+      else if (post.author.authorId != user.id)
+        return new httpErrors.Forbidden();
+      var body = req.body;
+
+      // ExpireAt needs to calculate the date
+      if ("expireAt" in body) {
+        switch (body.expireAt) {
+          case "day":
+            body.expireAt = (1).days().fromNow();
+            break;
+          case "week":
+            body.expireAt = (7).days().fromNow();
+            break;
+          case "month":
+            body.expireAt = (1).months().fromNow();
+            break;
         }
-      });
-      return post.save();
+      }
+
+      return Object.assign(post, body).save();
```

Comment:
The point of having `if (post[key] && post[key] !== req.body[key])` was some additional security to prevent this request from simply adding anything coming out of the request.body and overwriting it directly... 
I suppose there could be some edge cases here but I think the schena's should take care of validating the request body so except for some data transformations this could work... but it seems like error-prone to me
The `post` object here is an instance of a Mongo document with prototype methods such as `.save()` so I think we should use something like `post.overwrite(body).save()` instead ( see https://mongoosejs.com/docs/documents.html )